src/core/address_payload_unification_design.md

```diff
@@ -0,0 +1,556 @@
+# Design: Address–payload unification
+
+The purpose of this document is to describe a system by which we can
+remove the hard separation between node addresses and payloads, as
+described in [sourcecred/sourcecred#190][issue-190].
+
+[issue-190]: https://github.com/sourcecred/sourcecred/issues/190
+
+This document typechecks—run its source through [`print_code_fences`]
+and copy the result into https://flow.org/try.
+
+[`print_code_fences`]: https://gist.github.com/wchargin/e371d4f3047cf5d89e7ff9bb47224243
+
+This document was written at 2b301f91597a4502612f53a62bdf0367b9249bd4;
+temporal references like “current” and “unchanged” have this reference
+point.
+
+## Core structures
+
+The `Address` type is, for now, unchanged:
+
+```javascript
+type Address = {|
+  +pluginName: string,
+  +type: string,
+  +id: string,
+|};
+```
+
+At the core are two interfaces. First is `NodePayload`:
+
+```javascript
+interface NodePayload {
+  address(): Address;
+}
+```
+
+Plugins are expected to have a variety of subclasses of `NodePayload`,
+one for each logical type. These classes can provide domain-specific
+porcelain. For instance, the GitHub plugin might have:
+
+```javascript
+type GithubNodeType = "PULL_REQUEST" | "PULL_REQUEST_COMMENT";  // e.g.
+class GithubNodePayload<+T: {+url: string}> implements NodePayload {
+  +record: T;
+  +_type: GithubNodeType;
+  constructor(record: T, type: GithubNodeType) {
+    this.record = record;
+    this._type = type;
+  }
+  address() {
+    return {
+      pluginName: "sourcecred/github-beta",
+      type: this._type,
+      id: this.url(),
+    };
+  }
+  url() {
+    return this.record.url;
+  }
+  toJSON() { throw new Error("Must be implemented by subclass."); }
+  static fromJSON() { throw new Error("Must be implemented by subclass."); }
+}
+
+type PullRequestJson = {|
+  +url: string,
+  +number: number,
+  +title: string,
+  +body: string,
+|};
+class PullRequestPayload extends GithubNodePayload<PullRequestJson> {
+  constructor(record: PullRequestJson) {
+    super(record, "PULL_REQUEST");
+  }
+  number(): number { return this.record.number; }
+  title(): string { return this.record.title; }
+  body(): string { return this.record.body; }
+  toJSON() { return toCompat(PULL_REQUEST_COMPAT_INFO, this.record); }
+  static fromJSON(json) { return fromCompat(PULL_REQUEST_COMPAT_INFO, json); }
+}
+const PULL_REQUEST_COMPAT_INFO: CompatInfo = {
+  type: "sourcecred/sourcecred/github/PullRequest",
```

Comment:
Hmm, when we serialize the Graph JSON this will be considerable bloat, and is wholly redundant with the `pluginName` and `type` fields. Can we think of a way to deduplicate this with the pluginName/type, or to use index references rather than including the full string every time?

If we internally store nodes in a `pluginName/type` hierarchy, then potentially we could apply the compat_info to an array of all the payloads rather than on each one; this would also make it easy to deduplicate the address info.

Also, I know you want to de-privilege the "type" field, but notice that this is an orthogonal case where we have a first-class `type` construct. More to say on this in the part of the doc where you propose de-privileging type.

Reply:
Interesting point with the splitting-author case. I'm doubtful that we'd want a single `CompatInfo` for each of those node types; at the time when we split into three distinct types, they should have three `CompatInfos`, which reflects the fat that they can version independently.

It's not quite clear what to do with the legacy `Author` nodes; it seems to suggest that maybe the `CompatInfo` should be at the level of the plugin, not the individual node. Then the plugin can define compat handlers so that pre-split plugin data can be properly reconciled when it's loaded by post-split handlers.